cirq/ops/linear_combinations.py

```diff
@@ -205,3 +207,54 @@ def extend(expansion: value.LinearDict[str],
             extended_expansion = extend(expansion, op.qubits, self.qubits)
             result += extended_expansion * coefficient
         return result
+
+
+class PauliSum:
+    """Represents operator defined by linear combination of PauliStrings.
+
+    Since PauliStrings store their own coefficients, this class
+    does not implement the LinearDict interface. Instead, you can
+    add and subtract terms and then iterate over the resulting
+    (simplified) expression.
+
+    Under the hood, this class is backed by a LinearDict with coefficient-less
+    PauliStrings as keys. PauliStrings are reconstructed on-the-fly during
+    iteration.
+
+    """
+
+    def __init__(self, linear_dict):
+        self._linear_dict = linear_dict
+
+    @classmethod
+    def from_pauli_strings(cls, terms: List[PauliString]):
+        terms = {pstring.unit(): pstring.coefficient for pstring in terms}
+        return cls(linear_dict=value.LinearDict(terms))
+
+    def __iter__(self):
+        for vec, coeff in self._linear_dict.items():
+            yield PauliString.from_unit(vec, coeff)
+
+    def __iadd__(self, other):
+        if isinstance(other, PauliString):
+            other = PauliSum.from_pauli_strings([other])
+
+        self._linear_dict += other._linear_dict
+        return self
+
+    def __isub__(self, other):
+        if isinstance(other, PauliString):
+            other = PauliSum.from_pauli_strings([other])
+        self._linear_dict -= other
+        return self
+
+    def __repr__(self) -> str:
```

Comment:
Implement `__eq__`, `__hash__`, `_approx_equal_`, and the new `_equal_up_to_global_phase_` (may not be merged yet). Note that you can use `@value.value_equality(approximate=true)` for the first three.